### PR TITLE
meta: migrate TransactionQueries to TS

### PR DESCRIPTION
**Pull request by WikiRik**
## Pull Request Checklist

- [ ] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [ ] Did you update the typescript typings accordingly (if applicable)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [ ] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

I took a shot at migrating TransactionQueries to TypeScript, in order to write some unit tests for its functions.
I came across three things I'm definitely not sure of;
1. `setIsolationLevelQuery` and `commitTransactionQuery` returned both void and string. But that's not a consistent return, is it fine to return `''` instead? Or do we allow both to be returned in the typings and disable the consistent-return rule of eslint?
2. Some functions referenced the `quoteIdentifier` and `dialect` from the QueryGenerator. How can I use those here?
3. In Transaction `name` and `parent` are private, but they are used in this file so I made them public. Is that the way to go or do we want to tackle this differently?

## Todos

- [ ] 
- [ ] 
- [ ]

### CODE REVIEW

**Comment by WikiRik** [comment]
> `setIsolationLevelQuery` and `commitTransactionQuery` should always return the actual query, but there is a design issue with the current implementation. `setIsolationLevel`, `commitTransaction` shouldn't be called if the "transaction" is really a savepoint.

That makes sense. Do we want the type of these QG functions to be a modified version of Transaction without the parent property?

> `startTransactionQuery` shouldn't return a savepoint query if the transaction is a savepoint, it should throw. We should have a separate `createSavePointQuery`
> `name` should remain private. `createSavePointQuery` should request the save point ID instead of a Transaction object

I'll take a look at these
 
> by the way `query-generator/transaction.ts` must go away, these methods need to go in `AbstractQueryGeneratorTypeScript` instead. We can't use mixins with typescript

I was already afraid so, I'll update it.

**Comment by ephys** [comment]
I prefer to avoid migrating something to TypeScript until its API can be considered stable & public. Here is everything I've found to get to that goal:

queryGenerator:

- setIsolationLevelQuery: should only take value parameter, no options.
- startTransactionQuery:
  - Should not accept a transaction but an option bag.
  - The option bag is only used by sqlite (for "type"). Other dialects also support options in START TRANSACTION, but our implementation is inconsistant there. This should maybe be deferred to its own PR so we can overhaul it.
  - save points should not use `startTransactionQuery` but a separate "createSavePointQuery"
- createSavePointQuery: new method. Accepts a single "savepointName" parameter
- deferConstraintsQuery: 
  - looks completely broken? Used by `queryInterface#deferConstraints` but the SQL it produces isn't valid at all. We have 0 tests about this, or the query interface method, or Transaction.setDefferable.
  - The source of the issue comes from that fact that this method accepts a Deferrable, but that the Deferrable class is used for both marking attributes as deferrable (Deferrable.INITIALLY_DEFERRED, Deferrable.INITIALLY_IMMEDIATE, Deferrable.NOT), and setting the constraint deferrability for the current transaction (Deferrable.SET_DEFERRED, Deferrable.SET_IMMEDIATE). These are two completely different concepts that must be split.
    This explains why `deferConstraintsQuery` is used by both `queryInterface#deferConstraints` (in which case it must produce a compelte query, and use the latter two defferable values), and `queryGenerator#getConstraintSnippet` (in which case it produces only part of a query, and uses the first 3 deferrable values).
  - highly recommend leaving it in the old file and handling it in its own PR.
- setConstraintQuery:
  - looks like an internal method, move to PostgresQueryGenerator and make it a private method (hard private, with #).
- setDeferredQuery:
  - rename to "setConstraintsDeferredQuery"
  - "columns" parameters is actually "constraintNames" (string array). It's optional
  - sounds like only postgres supports this. Dialect that support it could throw "is not implemented" while dialect that plainly don't support it should throw "is not supported by x"
- setImmediateQuery:
  - same as above but renamed to "setConstraintsImmediateQuery"
- commitTransactionQuery
  - does not need a parameter. This method shouldn't be called by savepoints instead of this method aborting
- rollbackTransactionQuery
  - save point logic must also be separated from this method by adding `rollbackSavepointQuery`

queryInterface:

- setIsolationLevel:
  - First parameter should not be a "transaction" object, that option is already present in the option bag. It should also call setTransactionFromAls on the option bag at the very start of the method.
  - setIsolationLevel should not worry about whether the transaction is a savepoint. But the `Transaction#begin` should not attempt to change the isolation level for savepoints, so setIsolationLevel should not be called at all for savepoints. If the user tries to call queryInterface.setIsolationLevel manually, we should let the DB throw an error.
- startTransaction:
  - This method should be split in two: `createSavepoint` and `startTransaction`
  - Users should not attempt to use this method directly, so it should not be part of our public API.
  - Maybe we should split QueryInterface into "QueryInterface" (public API) and "InternalQueryInterface" (only accessible by sequelize's internals).
- deferConstraints: Similarly to deferConstraintsQuery: leave as-is for now. Handle in own PR.
- commitTransaction, rollbackTransaction: Same as startTransaction: internal APIs. Using them directly is always going to lead to problems.

---

I've also noticed something else while researching this:

In DB2, if you nest transactions, it doesn't create a savepoint. Instead it commits the previous transaction and starts a new one. This is not what it should do, especially since it supports savepoints https://www.ibm.com/docs/en/db2/10.1.0?topic=statements-savepoint.
We should add a test to ensure this is done properly (in its own PR).

**Comment by WikiRik** [comment]
Okay, to recap:

This PR will migrate the following to AbstractQueryGeneratorTypeScript;
- setIsolationLevelQuery
- startTransactionQuery (with a TODO for the type option)
- createSavepointQuery
- setDeferredQuery (renamed to setConstraintsDeferredQuery)
- setImmediateQuery (renamed to setConstraintsImmediateQuery)
- rollbackTransactionQuery
- rollbackSavepointQuery

And move the following to PostgresQueryGeneratorTypeScript;
- setConstraintQuery

It will also update the `startTransaction` and `rollbackTransaction` functions in QueryInterface a little bit for now in order to deal with the new savepoint queries. They will be migrated to TS in a future PR.

---

Future PRs will look into:
- type options in `startTransactionQuery` (also in relation to other dialects)
- deferConstraints and deferConstraintsQuery
- setIsolationLevel
- startTransaction, commitTransaction, rollbackTransaction
- how to deal with internal QueryInterface functions
- savepoints in nested transactions in DB2

File: src/transaction.ts [hunk 1]
```diff
@@ -19,9 +19,10 @@ export class Transaction {
   private readonly _afterCommitHooks: Set<AfterTransactionCommitCallback> = new Set();
   private readonly savepoints: Transaction[] = [];
   private readonly options: PartlyRequired<TransactionOptions, 'type' | 'isolationLevel' | 'readOnly'>;
-  private readonly parent: Transaction | null;
+  readonly parent: Transaction | null;
   readonly id: string;
-  private readonly name: string;
+  // TODO: this should become private, the savepoint queries should request the save point ID instead of a Transaction object
```
Comment (WikiRik) [inline src/transaction.ts:24]: I wasn't sure, but is the savepoint ID just the Transaction id?

File: src/dialects/db2/query-generator.js [hunk 2]
```diff
@@ -857,36 +857,40 @@ export class Db2QueryGenerator extends Db2QueryGeneratorTypeScript {
     return '';
   }
 
+  // TODO: migrate to Db2QueryGeneratorTypeScript
+  // TODO: how to deal with this void return? since the abstract method returns string
```
Comment (WikiRik) [inline src/dialects/db2/query-generator.js:861]: I tried returning `string | void` in the abstract method, but eslint complains due to https://typescript-eslint.io/rules/no-invalid-void-type/

**Review by WikiRik — COMMENTED** [review]
@ephys I have two specific questions for you, see my comments

File: src/dialects/db2/query-generator.js (hunk 2 above)
Comment (ephys) [reply src/dialects/db2/query-generator.js:861]: Something that doesn't support setting the isolation level should throw if someone tries to use that method

File: src/transaction.ts (hunk 1 above)
Comment (WikiRik) [reply src/transaction.ts:24]: This one?
https://github.com/sequelize/sequelize/blob/eb0ecbe44afb2284a18595360d653f2b68c3cc5e/src/transaction.ts#L58

File: src/dialects/db2/query-generator.js (hunk 2 above)
Comment (WikiRik) [reply src/dialects/db2/query-generator.js:861]: Oh yeah, that should also works. Forgot about that option

File: src/transaction.ts (hunk 1 above)
Comment (ephys) [reply src/transaction.ts]: Yep

**Comment by WikiRik** [comment]
Since we're focusing on getting the v7 beta ready this one will most likely become stale for a while, just fyi

**Comment by WikiRik** [comment]
Part of this is already done by #15962 so I'll close this in favour of that PR and we can revisit what remains after

**Pull request closed without merging**